allow grouping multiple view functions into one elm file

folkertdev opened the pull request:
**my use case** 

I would like to use this library to generate a bunch of html templates (for hakyll). They all depend on 
the same style code, so in most cases, I want to regenerate all my templates when I change some elm code. 

Currently, putting all these view functions (about 10) in a .js file and calling this lib for all of them is too slow (5 seconds), because 10 elm files are generated and compiled (it also seems that my code
as it is imported, is compiled multiple times). 

**proposed fix**

Allow for putting multiple view functions together, compiling them into one elm file. This PR introduces 

```typescript
export interface ViewFunctionConfig {
    viewFunction: string;
    output: string;
    model?: any;
    decoder?: string;
    indent?: number;
    newLines?: boolean;
}

export function grouped(
    rootDir: string, moduleName: string, configs: ViewFunctionConfig[],
    alreadyRun?: boolean, elmMakePath?: string, installMethod?: string): Promise<void[]> 
```

*the moduleName argument is only used to generate a unique hash. Might be better to concatenate the view function names and take that as the hash)*

This function does the writing to file for you, because otherwise the user needs to keep track of the order of the view functions and then reassociate the rendered html with the correct output file later.

Alternatively we could still take the filename as an input but give back a `Promise<[[string, string]]>` where the first element is the rendered html and the other is the file name. 

<hr> 

Does this seem good architecturally? the PR actually contains code that implements this idea but it's not very pretty yet and I'd like to check whether this is the right idea first.

folkertdev committed:
Adding multiple views/outputs per elm file

This PR adds support for grouping multiple views/output files.

This is useful when many of your views share depenencies: these will now
be compiled only once. This is a double-edged sword because when you
want to regenerate one of the views all the code has to be compiled.

Added is the function `elmStaticHtmlMultiple`
```typescript
elmStaticHtmlMultiple(
    rootDir: string,
    moduleName: string,
    configs: ViewFunctionConfig[],
    alreadyRun?: boolean,
    elmMakePath?: string,
    installMethod?: string
    ): Promise<void[]>
```

calling works like

```typescript
const configs =
    [ { viewFunction: "MyModule.view", model: model
        , decoder: "MyModule.decodeModel", output: "multiple1.html" }
        , { viewFunction: "MyModule.lazyView", model: model
        , decoder: "MyModule.decodeModel", output: "multiple2.html" }
    ];

elmStaticHtmlMultiple(process.cwd(), "MyModule", configs);
```

A major difference with `elmStaticHtml` is that the output file is part
of the config and consequently a `Promise<void>` is returned.

The module name (2nd argument) is only there to create a unique hash

I'm sure there is a bunch of style things that need to be changed, but
does the basic architecture make sense?

folkertdev committed:
rename to grouped

eeue56 commented on index.ts:
```diff
@@ -36,11 +36,11 @@ function parseProjectName(repoName: string): string {
     return new Promise((resolve, reject) => {
         const Elm = require(path.join(dirPath, "elm.js"));
-        const privateName = `PrivateMain${viewHash}`;
+        const privateName = `PrivateMain${moduleHash}`;
```
💭 why this change?

folkertdev replied:
we're combining multiple view functions into one file, so just picking the first view function to generate the hash seems arbitrary. The name would also clash if the same view function is used elsewhere.

eeue56 replied:
Gotcha

eeue56 commented on example/index.ts:
```diff
@@ -44,4 +48,18 @@ function runLazyView() {
+        [ { viewFunction: "MyModule.view", model: model
+            , decoder: "MyModule.decodeModel", output: "multiple1.html" } 
+            , { viewFunction: "MyModule.lazyView", model: model
+            , decoder: "MyModule.decodeModel", output: "multiple2.html" } 
```
💭 is this the preferred API for you? do you think there could be something better?

folkertdev replied:
I'm not sure, I agree it doesn't feel very nice.

It depends on what one wants to do with the files, I guess. For me, I just want to save them to disk, so the above makes sense. It's possible to return a `Promise<[string]>`, then the user 
has to re-associate the generated html and the file name:

```typescript
function runMultiple() {
    const configs =
        [ { viewFunction: "MyModule.view", model, decoder: "MyModule.decodeModel", output: "multiple1.html" }
        , { viewFunction: "MyModule.lazyView", model, decoder: "MyModule.decodeModel", output: "multiple2.html" }
        ];

    elmStaticHtml.grouped(process.cwd(), "MyModule", configs)
        .then((generatedHtmls) => {
            generatedHtmls
                .map((generatedHtml, i) => fs.writeFileSync(configs[i].output, generatedHtml));
        });
}
```

folkertdev committed:
Make grouped return a Promise<[string]>

Also removing the `output` field from the config for grouped.

folkertdev committed:
Only compile files required for a step

This fixes a problem of clashing port names when a view function
is used multiple times in a grouped and a normal call of elmStaticHtml

folkertdev committed:
also hash the produced elm.js file

This seems to be needed because of asynchronicity. Removing it
will cause the `elm.js` file to be overwritten (by the next view/model
pair) before it can be run to produce output.

Normally we might be able to compile multiple apps into one elm.js file,
but the port namespace still seems to be global and port names are
linked to view names, so if the same view function is used multiple
times, this will cause an error. (clash in port names).

folkertdev commented:
I've switched to returning a `Promise<[string]>`. The extra flexibility of having a list of strings
to manipulate is probably worth it. 

To make grouped work together with other elmStaticHtml calls, some further changes were needed.

It seems that elm's port namespace is global, so even though we can compile multiple `main` functions into one javascript file, we can't have multiple ports with the same name. 

There are then two solutions: only generate the ports once (storing them in some `Ports.elm` file, or generating multiple .js files. The second option is simpler and thus the one I went with for now. 
I vaguely remember hearing something about this problem, so maybe this issue is solved in 0.19?

folkertdev committed:
Revert "also hash the produced elm.js file"

This reverts commit 978b5cb6162e692bce12f7a979175f703ccec7cd.

folkertdev committed:
Revert "Only compile files required for a step"

This reverts commit 8e8b8f27efbe15af72aae222e10a6284360b1baa.

eeue56 commented on example/index.ts:
```diff
@@ -24,24 +25,40 @@ function runTwice() {
+    elmStaticHtml.grouped(process.cwd(), "MyModule", configs)
+        .then((generatedHtmls) => {
+            generatedHtmls
+                .map((generatedHtml, i) => fs.writeFileSync(configs[i].output, generatedHtml));
```
Relying on order kinda sucks: why not make make it so that the returned stuff is a list like this:

```
List { generatedHtml : String,  fileOutputName: String }
```
?

eeue56 commented on index.ts:
```diff
@@ -82,13 +82,77 @@ function makeHash(viewFunction: string): string {
+
+// compiles multiple view functions into one elm file
+// which is much faster if you're likely to need all of them
+export function grouped(
```
🎨  the name `grouped` doesn't make much sense to me, though I know why it's called that. How about something like `elmStaticHtml.multiple()` instead?

eeue56 commented on index.ts:
```diff
@@ -36,11 +36,11 @@ function parseProjectName(repoName: string): string {
 }
 
-function runElmApp(viewHash: string, dirPath: string, model: any): Promise<string> {
+function runElmApp(moduleHash: string, viewHash: string, dirPath: string, model: any): Promise<string> {
```
🎨 what's the purpose of the viewHash if we're using a module hash instead? Shouldn't it be a combination of moduleName + viewName?

eeue56 commented on index.ts:
```diff
@@ -82,13 +82,77 @@ function makeHash(viewFunction: string): string {
+// which is much faster if you're likely to need all of them
+export function grouped(
+    rootDir: string, moduleName: string, configs: ViewFunctionConfig[],
+    alreadyRun?: boolean, elmMakePath?: string, installMethod?: string): Promise<string[]> {
```
💭 I wonder if returning `Promise<string, string>[]` would be better: that way it wouldn't block on all of them finishing

folkertdev replied:
Does this have any benefit when the elm app isn't truly parallel? 

You could send in the input via ports, which in theory could give parallelism, but elm doesn't currently support that.  Instead you could spawn the elm app multiple times, but that probably comes with considerable overhead.

But maybe when you have enough models with a small number of view functions, that's worth it?

eeue56 commented on index.ts:
```diff
@@ -82,13 +82,77 @@ function makeHash(viewFunction: string): string {
 
     if (options.alreadyRun === true) {
-        return runElmApp(viewHash, dirPath, options.model);
+        return runElmApp(viewHash, viewHash, dirPath, options.model);
```
🤔 Why give the `viewHash` twice?

eeue56 commented on index.ts:
```diff
@@ -82,13 +82,77 @@ function makeHash(viewFunction: string): string {
+export function grouped(
+    rootDir: string, moduleName: string, configs: ViewFunctionConfig[],
+    alreadyRun?: boolean, elmMakePath?: string, installMethod?: string): Promise<string[]> {
+    const moduleHash = makeHash(moduleName);
```
💭 why does it need a module hash? Why do you need to say which module the functions are in? Don't we already know the module name from the `viewFunction`?

eeue56 commented on index.ts:
```diff
@@ -160,7 +224,45 @@ function runCompiler(viewHash: string,
+                         rootDir: string, configs: ViewFunctionConfig[], elmMakePath?: string): Promise<string[]> {
+    const options: any = {
+        cwd: rootDir,
+        output: `elm${moduleHash}.js`,
```
💭 I think this should be a hash of all the viewFunctions joined

folkertdev commented:
I've started work on the single-port approach we discussed on slack. It makes some of the problems 
with hashes you just highlighted go away and cleans up the code considerably. 

I hope to have that working and incorporate the above comments by saturday. Thanks for the review

eeue56 commented:
Thanks for the PR!

folkertdev committed:
make default use the grouped code under the hood

folkertdev committed:
rename grouped to multiple

folkertdev committed:
thread the filename through

folkertdev committed:
return a { generatedHtml, fileOutputName } record

Friendlier than a 2-element array. Also fixes a bug where
the model was not defined in a config.

folkertdev committed:
switch to fileOutputName everywhere

folkertdev commented:
found some time today

The above commits

* implement the "one port per generated file" idea.
* remove duplication (`elmStaticHtml` is seens as a singleton `multiple`)
* fix issues with hashes
* thread the `fileOutputName` through

One further idea: compile into separate .js files. This diff gives a 30% speed increase (5.1 to 3.4 seconds) on my machine on `examples/index.ts`. Including all the `PrivateMain` files results in n! compilations, compiling just to the PrivateMain needed is constant.

```diff
diff --git a/index.ts b/index.ts
index 7a3e49f..0a2228f 100644
--- a/index.ts
+++ b/index.ts
@@ -39,7 +39,7 @@ function parseProjectName(repoName: string): string {
 function runElmApp(moduleHash: string, dirPath: string, filenamesAndModels: any[][]): Promise<Output[]> {
 
     return new Promise((resolve, reject) => {
-        const Elm = require(path.join(dirPath, "elm.js"));
+        const Elm = require(path.join(dirPath, `elm${moduleHash}.js`));^M
         const privateName = `PrivateMain${moduleHash}`;
 
         if (Object.keys(Elm).indexOf(privateName) === - 1) {
@@ -218,7 +218,7 @@ function runCompiler(moduleHash: string,
                      configs: ViewFunctionConfig[], elmMakePath?: string): Promise<Output[]> {
     const options: any = {
         cwd: rootDir,
-        output: "elm.js",
+        output: `elm${moduleHash}.js`,^M
         yes: true,
     };
 
@@ -228,7 +228,7 @@ function runCompiler(moduleHash: string,
 
     return new Promise((resolve, reject) => {
         fs.readdir(rootDir, (err, files) => {
-            const actualFiles = files.filter((name) => name.indexOf("PrivateMain") === 0);
+            const actualFiles = files.filter((name) => name.indexOf(`PrivateMain${moduleHash}`) === 0);^M
 
             const compileProcess = compile(actualFiles, options);
             compileProcess.on("exit",

```

This means more files are generated (and the cache removal needs to be a bit more complex) but for 30% speed gain it seems worthwhile. thoughts?

eeue56 commented:
> One further idea: compile into separate .js files. This diff gives a 30% speed increase (5.1 to 3.4 seconds) on my machine on examples/index.ts. Including all the PrivateMain files results in n! compilations, compiling just to the PrivateMain needed is constant.

Why is this faster? 🤔

eeue56 commented on index.ts:
```diff
@@ -82,13 +76,94 @@ function makeHash(viewFunction: string): string {
+        , model: options.model
+        , newLines: options.newLines
+        , viewFunction
+        , viewHash};
```
Not a fan of this style of commas in JS, best to put them afterwards 🎨

eeue56 commented on templates.ts:
```diff
@@ -75,41 +49,134 @@ export function generateRendererFile(
+function generateBody(config: ViewFunctionConfig): string {
+    const optionsSet = generateOptionsSet(config.newLines, config.indent);
+    if (config.decoder) {
+        console.log("withDecoder ->", config.decoder, config.viewFunction);
```
is this for debugging? if so, make it an optional thing. i used `HIDE_WARNINGS` for this on master

eeue56 commented on templates.ts:
```diff
@@ -75,41 +49,134 @@ export function generateRendererFile(
+        .map((config) => `render${config.viewHash}`)
+            .join(", ");
+
+    const port = `port htmlOut${hash} : List { generatedHtml : String, fileOutputName: String } -> Cmd msg`;
```
🎉 much better

eeue56 commented on example/index.ts:
```diff
@@ -24,24 +23,38 @@ function runTwice() {
+    elmStaticHtml.multiple(process.cwd(), configs)
+        .then((generatedHtmls) => {
+            generatedHtmls
+                .map((output) => fs.writeFileSync(output.fileOutputName, output.generatedHtml));
```
🎉 much better

eeue56 commented on index.ts:
```diff
@@ -82,13 +76,94 @@ function makeHash(viewFunction: string): string {
+
+export interface Output {
+    generatedHtml: string;
+    fileOutputName: string;
```
this should be documented in the readme too

eeue56 commented on index.ts:
```diff
@@ -82,13 +76,94 @@ function makeHash(viewFunction: string): string {
+
+// compiles multiple view functions into one elm file
+// which is much faster if you're likely to need all of them
+export function multiple(
```
usage of this should be documented in the readme

eeue56 commented on example/index.ts:
```diff
@@ -1,20 +1,19 @@
         .then((generatedHtml) => {
             fs.writeFileSync("output2.html", generatedHtml);
-            elmStaticHtml(process.cwd(), "MyModule.view", secondRunOptions)
+            elmStaticHtml.default(process.cwd(), "MyModule.view", secondRunOptions)
```
This usage seems wrong,based on the changes you made. Doesn't it return an object now, like the multiple one? I'd rather it didn't return an object if it's a single view

folkertdev replied:
struggling a bit with typescripts module system. 

I think `elmStaticHtml` can't be the default and also be a module. Would it be OK to drop the default and just expose `elmStaticHtml` and `multiple`, and then import as 

```typescript
import { elmStaticHtml, multiple } from "elm-static-html-lib";
```

folkertdev commented:
The compilation is slow because elm files are needlessly compiled multiple times.

say we will generate 3 privatemain files

* the first file is generated, elm.js is created by compiling all PrivateMain*.elm files, which is only one.
* the second file is generated, again elm.js is created by compiling all PrivateMain*.elm files, now two files
* the third file is generated, again elm.js is created by compiling all PrivateMain*.elm files, now 3 files

So 3 elm files results in 1 + 2 + 3 compilations (this is n^2 complexity). Ideally, you'd want to only generate JS once all the elm is generated, but that is impossible to enforce with the singular `elmStaticHtml` commands.

`grouped` works around this problem by only generating one `PrivateMain` for multiple views, but we can solve the problem in general by generating a `elm.js` file per elm file.

eeue56 commented:
Gotcha. I think we should optimize for runtime, rather than compile time. Maybe that can be a different PR later on?

folkertdev commented:
are multiple applications compiled into one file faster than separate applications? Otherwise there would only be more disk space used, the runtime is unchanged.

but yes this can be a separate thing.

eeue56 commented:
> are multiple applications compiled into one file faster than separate applications? Otherwise there would only be more disk space used, the runtime is unchanged. 

Yep -- requiring a single file in the JS side is faster than requiring as needed

folkertdev committed:
add to README

folkertdev commented:
* fixed the style issues (console.log use and commas)
* As mentioned above, I think the only way to expose the two functions is to not have a default function, but maybe there is some importing trick that makes it possible that I'm unaware of. 
* Added the new function and types to the readme, Not sure whether the typescript types should be documented as such or as their pure JS versions. I think it's best you make a pass over the readme too to ensure consistent style.

folkertdev committed:
rework exports/imports

folkertdev committed:
remove debug console.logs

eeue56 commented on example/index.ts:
```diff
@@ -1,12 +1,11 @@
-import elmStaticHtml from "../index";
 import * as fs from "fs";
+import { elmStaticHtml, multiple } from "../index";
```
👍  I'm good with this

eeue56 replied:
Though maybe it could be `single`, but I'm okay with this

eeue56 commented:
Published as 0.1.0! Thanks for your hard work 🎉
